### PR TITLE
VAULT-32804: Add STS Fallback parameters to secrets-aws engine

### PR DESCRIPTION
**Pull request by kpcraig**
### Description
Adds two fields, `sts_fallback_endpoints` and `sts_fallback_regions` to allow credential requests to try other sts regions if the first one fails.

To test this one, set up a `secrets/aws` mount in the usual way:

```
vault auth enable aws
```

```
vault write aws/config/root \
  access_key=$AWS_ACCESS_KEY_ID secret_key=$AWS_SECRET_ACCESS_KEY \
  sts_endpoint=https://sts.us-west-1.amazonaws.com
  sts_region=us-west-1
  sts_fallback_endpoints=https://sts.us-east-2.amazonaws.com,https://sts.us-east-1.amazonaws.com
  sts_fallback_regions=us-east-2,us-east-1
```

```
vault write aws/roles/dev-role-iam credential_type=assumed_role role_arns=arn:aws:iam::XXXXXXXXXX:role/foobar
```

```
vault read aws/sts/dev-role-iam

> Key                Value
---                -----
lease_id           aws/creds/dev-role-iam/.....
```

Now, 'break' us-west-1 somehow, (I suggest messing with `/etc/hosts`), and you should get a failure:

```
vault read aws/sts/dev-role-iam

Error reading aws/creds/dev-role-iam: Error making API request.

URL: GET http://localhost:8200/v1/aws/creds/dev-role-iam
Code: 400. Errors:

* Error assuming role: RequestError: send request failed
caused by: Post "https://sts.us-east-2.amazonaws.com/": dial tcp 127.0.0.1:443: connect: connection refused
```

Experiment with breaking and unbreaking the route to AWS servers - the credential should succeed as long as one sts domain works.

### TODO only if you're a HashiCorp employee
- [X] **Jira:** If this change has an associated Jira, it's referenced either
  in the PR description, commit message, or branch name.

### CODE REVIEW

File: builtin/logical/aws/path_config_root_test.go
Comment (fairclothjm) [inline builtin/logical/aws/path_config_root_test.go:35]: Are there more test cases we can add for this?

File: builtin/logical/aws/client.go
Comment (robmonte) [inline builtin/logical/aws/client.go:158]: Is this case only likely to happen when setting the new fallback fields while using IAM credentials instead of STS?

Comment (kpcraig) [reply builtin/logical/aws/client.go:158]: the fields should only get appended to in the case of using STS client type, so theoretically this case should never happen

**Review by robmonte — APPROVED** [review]

File: builtin/logical/aws/path_config_root_test.go
Comment (kpcraig) [reply builtin/logical/aws/path_config_root_test.go:35]: added a couple tests cases - normal input, comma separated string, and mismatched entry error check.

File: builtin/logical/aws/client.go
Comment (fairclothjm) [inline builtin/logical/aws/client.go:29]: suggest changing to plural since we are returning an array
```suggestion
func (b *backend) getRootConfigs(ctx context.Context, s logical.Storage, clientType string, logger hclog.Logger) ([]*aws.Config, error) {
```

File: builtin/logical/aws/client.go
Comment (fairclothjm) [inline builtin/logical/aws/client.go:47]: I believe this is possible since none of the `aws/config/root` fields are required by the API as they can be either set to defaults or the environment. i.e., I don't believe it is required to explicitly call `aws/config/root` and in that case entry will be nil.

File: builtin/logical/aws/client.go
Comment (fairclothjm) [inline builtin/logical/aws/client.go:193]: Should we log what region and endpoint failed?

**Review by fairclothjm — DISMISSED** [review]
Overall, LGTM! I added a few suggestions but they should not hold this up.

File: builtin/logical/aws/client.go
Comment (kpcraig) [reply builtin/logical/aws/client.go:193]: good idea

File: builtin/logical/aws/client.go
Comment (kpcraig) [reply builtin/logical/aws/client.go]: I've updated the comment to reflect this

File: builtin/logical/aws/client.go
Comment (kpcraig) [reply builtin/logical/aws/client.go]: done!

**Review by fairclothjm — DISMISSED** [review]

**Review by schavis — APPROVED** [review]
Please apply the formatting fix before merging. Otherwise lgtm

**Review by robmonte — APPROVED** [review]

**Review by schavis — APPROVED** [review]

**Pull request merged**
